### PR TITLE
feat(`backbones`): Add different popular backbones

### PR DESCRIPTION
**Pull request by mateuuszzzzz**
PR was mainly created to add different backbones but during development some problems were discovered so I addressed them as well.

# Changes:
#### 5. Added query parameter that fixes problem with empty evaluation sets.

### CODE REVIEW

File: gwnmo/neptune_logger.py [hunk 1]
```diff
@@ -16,21 +16,27 @@ def __init__(self, online: bool):
         super(NeptuneLogger, self).__init__()
         self.output = self.print_to_term
         if online:
-            self.run = neptune.init_run()
+            try:
```
Comment (OneAndZero24) [inline gwnmo/neptune_logger.py]: why pass on except? fallback to offline maybe?

File: gwnmo/neptune_logger.py [hunk 2]
```diff
@@ -16,21 +16,27 @@ def __init__(self, online: bool):
...
+            except:
+                pass
             self.output = self.push_to_neptune
 
     def print_to_term(self, key, value):
         clock = time.strftime("%H:%M:%S", time.localtime(time.time()))
         print(f'[{clock}] {key}: {value}')
 
     def push_to_neptune(self, key, value):
+        try:
```
Comment (OneAndZero24) [inline gwnmo/neptune_logger.py]: after adding fallback to offline this try catch can be removed

File: gwnmo/neptune_logger.py (hunk 1 above)
Comment (mateuuszzzzz) [reply gwnmo/neptune_logger.py]: This was one thing that I want to delete. I had problems with Neptune so I used this to overcome this problem. I will delete all added exceptions catches

File: gwnmo/__main__.py [hunk 3]
```diff
@@ -30,8 +30,8 @@
         train(dataset_gen(), args.epochs, args.reps, module)
     else:
         if Module == GWNMOFS:
-            module = Module(args.lr, args.lr2, args.gamma, not args.nonorm, args.steps, args.ways, args.shots)
+            module = Module(args.lr, args.lr2, args.gamma, not args.nonorm, args.steps, args.ways, args.shots, trainable_fe = args.trainable_fe, feature_extractor_backbone=args.backbone_type)
```
Comment (OneAndZero24) [inline gwnmo/__main__.py:33]: Why pass last two as key=value. Pls either pass all as key value or none.

Comment (OneAndZero24) [inline gwnmo/__main__.py]: Why pass last two as key=value. Pls either pass all as key value or none.

File: gwnmo/models/backbone.py [hunk 4]
```diff
@@ -0,0 +1,580 @@
...
+        x_normalized = x.div(x_norm+ 0.00001)
+        if not self.class_wise_learnable_norm:
+            L_norm = torch.norm(self.L.weight.data, p=2, dim =1).unsqueeze(1).expand_as(self.L.weight.data)
+            self.L.weight.data = self.L.weight.data.div(L_norm + 0.00001)
+        cos_dist = self.L(x_normalized) #matrix product by forward function, but when using WeightNorm, this also multiply the cosine distance by a class-wise learnable norm, see the issue#4&8 in the github
+        scores = self.scale_factor* (cos_dist)
+
+        return scores
+
+class Flatten(nn.Module):
```
Comment (OneAndZero24) [inline gwnmo/models/backbone.py]: Can't we use https://pytorch.org/docs/stable/generated/torch.nn.Flatten.html ?

File: gwnmo/models/backbone.py [hunk 5]
```diff
@@ -0,0 +1,580 @@
...
+
+
+    def forward(self, x):
+        if self.weight.fast is not None and self.bias.fast is not None:
+            out = F.linear(x, self.weight.fast, self.bias.fast) #weight.fast (fast weight) is the temporaily adapted weight
+        else:
+            out = super(Linear_fw, self).forward(x)
+        return out
+
+class BLinear_fw(Linear_fw): #used in BHMAML to forward input with fast weight
```
Comment (OneAndZero24) [inline gwnmo/models/backbone.py]: Pls no Bayes 😿

File: gwnmo/models/backbone.py [hunk 6]
```diff
@@ -0,0 +1,580 @@
...
+    def forward(self, x):
+        out = F.relu(self.layer1(x))
+        out = F.relu(self.layer2(out))
+        out = F.relu(self.layer3(out))
+        out = out.view(out.size(0), -1)
+        return out
+
+
+# just to test the kernel hypothesis
+class BackboneKernel(nn.Module):
```
Comment (OneAndZero24) [inline gwnmo/models/backbone.py]: Do we need BackboneKernel? Pls lets keep this codebase minimal

File: gwnmo/modules/fewshot/gwnmofs.py [hunk 7]
```diff
@@ -17,25 +17,44 @@ class GWNMOFS(FSModuleABC):
...
+                 trainable_fe: bool = False, 
+                 feature_extractor_backbone = None
+        ):
         super(GWNMOFS, self).__init__()
 
+        self.MO = MetaOptimizer().to(device)
+        self.MO.train()
+
+        
+        if not trainable_fe:
```
Comment (OneAndZero24) [inline gwnmo/modules/fewshot/gwnmofs.py:38]: Move FE to parameter will be nicer if we want to try more

File: gwnmo/neptune_logger.py [hunk 8]
```diff
@@ -14,10 +14,11 @@ class NeptuneLogger(Logger):
 
     def __init__(self, online: bool):
         super(NeptuneLogger, self).__init__()
-        self.output = self.print_to_term
         if online:
             self.run = neptune.init_run()
             self.output = self.push_to_neptune
+        else:
```
Comment (OneAndZero24) [inline gwnmo/neptune_logger.py:20]: Why add the else over default value ?

File: gwnmo/utils.py [hunk 9]
```diff
@@ -48,6 +50,7 @@ def _setup_arg_parser():
     parser.add_argument('--lr', type=float, default=0.01, required=False, help='Learning rate')
     parser.add_argument("--gamma", type=float, default=0.01, required=False, help='Gamma')
     parser.add_argument('--nonorm', action='store_true', help='Normalization in GWNMO/HG in fs & classic')
+    parser.add_argument('--debugger', action='store_true', help='Indicates debug run that won\'t be logged to neptune (use this in launch.json configurations)')
```
Comment (OneAndZero24) [inline gwnmo/utils.py]: I don't know if this flag is needed if it only turns Neptune debugger off as we have the --offline flag

**Review by OneAndZero24 — COMMENTED** [review]
noice PR M8 🐤

File: gwnmo/models/backbone.py (hunk 5 above)
Comment (mateuuszzzzz) [reply gwnmo/models/backbone.py]: Hahahahaha

File: gwnmo/modules/fewshot/gwnmofs.py (hunk 7 above)
Comment (mateuuszzzzz) [reply gwnmo/modules/fewshot/gwnmofs.py:38]: I'll keep it the old way. If there will be need to experiment more with feature extractors then we can move it to the upper scope for better customization of flow

File: gwnmo/neptune_logger.py (hunk 8 above)
Comment (mateuuszzzzz) [reply gwnmo/neptune_logger.py:20]: I think it's more clear since we do not overwrite `self.output` variable

File: gwnmo/utils.py (hunk 9 above)
Comment (mateuuszzzzz) [reply gwnmo/utils.py]: Good point!

File: gwnmo/modules/fewshot/gwnmofs.py (hunk 7 above)
Comment (mateuuszzzzz) [reply gwnmo/modules/fewshot/gwnmofs.py:38]: For now we pass all parameters as `args.{key_name}`

**Review by OneAndZero24 — APPROVED** [review]

**Pull request merged**